### PR TITLE
[FLOC-4500] Get compute_instance_id from metadata service

### PR DESCRIPTION
**Pull request by wallrj**
Fixes: https://clusterhq.atlassian.net/browse/FLOC-4500

This adds an additional mechanism for looking up the local instance ID from the metadata service, if it is available.
- The metadata service is _not_ available on Rackspace - TCP connection attempts time out.
- The configdrive is available on all our tested Rackspace OS images.
- Configdrive _may_ be available on devstack VMs depending on:
  - `force_config_drive`  in `nova.conf` (http://docs.openstack.org/newton/config-reference/compute/config-options.html)
  - `config_drive` parameter to `POST /servers` API  (http://developer.openstack.org/api-ref/compute/?expanded=create-server-detail)

I've tested this on a devstack with `force_config_drive False` and a VM started with `nova boot --config-drive=false ...`

All but one of the Cinder tests pass (and one of the EBS tests which should be skipped on this platform).

The failing test is one which tries to simulate a connection timeout.
For that, I used  an unroutable `10.0.0.0` IP address, but on devstack this results in a connection error rather than a connection timeout....I guess because one of the routers is responding with ICMP error message.

```
(4500) ubuntu@richardw-xenial-1:~/flocker$ curl  -v http://10.0.0.0/
*   Trying 10.0.0.0...
* Immediate connect fail for 10.0.0.0: Network is unreachable
* Closing connection 0
curl: (7) Couldn't connect to server
```

vs

```
(4500) ubuntu@richardw-xenial-1:~/flocker$ curl  -v http://192.0.2.1/
*   Trying 192.0.2.1...
* connect to 192.0.2.1 port 80 failed: Network is unreachable
* Failed to connect to 192.0.2.1 port 80: Network is unreachable
* Closing connection 0
curl: (7) Failed to connect to 192.0.2.1 port 80: Network is unreachable

```

vs

```
(4500) [root@richardw-centos7-1 flocker]# curl --connect-timeout 5 -v  http://192.0.2.1/                                                                                                         
* About to connect() to 192.0.2.1 port 80 (#0)
*   Trying 192.0.2.1...
* Connection timed out after 5009 milliseconds
* Closing connection 0
curl: (28) Connection timed out after 5009 milliseconds

```

I'll try and figure out a better way of forcing  a timeout or do some mocking or something.

```
Traceback (most recent call last):
  File "/root/.venvs/4500/lib/python2.7/site-packages/twisted/internet/defer.py", line 149, in maybeDeferred
    result = f(*args, **kw)
  File "/root/.venvs/4500/lib/python2.7/site-packages/testtools/testcase.py", line 719, in _run_test_method
    return self._get_test_method()()
  File "/root/.venvs/4500/lib/python2.7/site-packages/eliot/testing.py", line 300, in wrapper
    return function(self, *args, **kwargs)
  File "/root/.venvs/4500/lib/python2.7/site-packages/eliot/testing.py", line 332, in wrapper
    return function(self, logger)
  File "/root/flocker/flocker/node/agents/functional/test_cinder.py", line 1219, in test_timeout_error
    "flocker:node:agents:blockdevice:openstack:"
  File "/root/.venvs/4500/lib/python2.7/site-packages/eliot/testing.py", line 56, in assertContainsFields
    test.assertEqual(messageSubset, fields)
  File "/root/.venvs/4500/lib/python2.7/site-packages/testtools/testcase.py", line 411, in assertEqual
    self.assertThat(observed, matcher, message)
  File "/root/.venvs/4500/lib/python2.7/site-packages/testtools/testcase.py", line 498, in assertThat
    raise mismatch_error
testtools.matchers._impl.MismatchError: !=:
reference = {'message_type': u'flocker:node:agents:blockdevice:openstack:compute_instance_id:metadataservice_connection_error'}
actual    = {'message_type': 'flocker:node:agents:blockdevice:openstack:compute_instance_id:metadataservice_connect_timeout'}


flocker.node.agents.functional.test_cinder.MetadataFromServiceTests.test_timeout_error
===============================================================================
[FAIL]
Traceback (most recent call last):
Failure: testtools.testresult.real._StringException: Empty attachments:
  twisted-eliot-log
  twisted-log

Traceback (most recent call last):
  File "/root/flocker/flocker/node/agents/test/test_ebs.py", line 266, in test_returns_device_name_list
    self.assertGreater(len(devices), 0)
  File "/root/.venvs/4500/lib/python2.7/site-packages/unittest2/case.py", line 1233, in assertGreater
    self.fail(self._formatMessage(msg, standardMsg))
  File "/root/.venvs/4500/lib/python2.7/site-packages/unittest2/case.py", line 690, in fail
    raise self.failureException(msg)
AssertionError: 0 not greater than 0


flocker.node.agents.test.test_ebs.FindAllocatedDeviceTests.test_returns_device_name_list
-------------------------------------------------------------------------------
Ran 467 tests in 420.435s

FAILED (skips=91, failures=2, successes=374)
```

### CODE REVIEW

File: flocker/node/agents/cinder.py [hunk 1]
```diff
@@ -561,7 +599,33 @@ def compute_instance_id(self):
...
+            lambda: metadata_from_service(
+                metadata_service_endpoint=self._metadata_service_endpoint
+            ),
+        ]
+        for checker in metadata_checkers:
+            metadata = checker()
+            if metadata:
+                return metadata["uuid"]
+        try:
+            result = self._compute_instance_id_by_ipaddress_match()
```
Comment (vchernoy) [inline flocker/node/agents/cinder.py]: maybe just `return` here?

File: flocker/node/agents/cinder.py [hunk 2]
```diff
@@ -110,6 +111,46 @@ def metadata_from_config_drive(config_drive_label=CONFIG_DRIVE_LABEL):
...
+        try:
+            return response.json()
+        except ValueError as e:
+            Message.new(
+                message_type=(
+                    u"flocker:node:agents:blockdevice:openstack:"
+                    u"compute_instance_id:metadata_file_not_json"),
+                error_message=unicode(e),
+            ).write()
+
```
Comment (vchernoy) [inline flocker/node/agents/cinder.py]: Is `None` to be returned here (for cases `not response.ok` or `ValueError` is caught?

File: flocker/node/agents/cinder.py [hunk 3]
```diff
@@ -561,7 +599,33 @@ def compute_instance_id(self):
...
-            raise UnknownInstanceID(self)
+
+    def compute_instance_id(self):
+        """
+        Attempt to retrieve node UUID from the metadata in a config drive or
+        from the metadata service.
+        Fall back to finging the ``ACTIVE`` Nova API server with an
+        intersection of the IPv4 and IPv6 addresses on this node.
+        """
+        metadata_checkers = [
```
Comment (vchernoy) [inline flocker/node/agents/cinder.py]: Do you expect more checkers?

I think two calls will look simpler than loop over vector of lambdas.

File: flocker/node/agents/cinder.py (hunk 2 above)
Comment (wallrj) [inline flocker/node/agents/cinder.py]: Agreed, but `None` is the default return value so I'll leave this for now.

File: flocker/node/agents/cinder.py (hunk 1 above)
Comment (wallrj) [inline flocker/node/agents/cinder.py]: I'll leave it.

**Pull request merged**